c/src/table.rs

```diff
@@ -0,0 +1,541 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The LanceDB Authors
+
+//! Complete Table API implementation for LanceDB C bindings
+//!
+//! This module provides all table operations using Arrow-only APIs,
+//! combining both simple and full table functionality.
+
+use std::ffi::CStr;
+use std::os::raw::c_char;
+use std::ptr;
+
+use arrow_array::{Array, RecordBatch, RecordBatchReader, StructArray};
+use arrow_schema::{ArrowError, Schema};
+use futures::TryStreamExt;
+use lancedb::query::{ExecutableQuery, QueryBase};
+
+use crate::connection::{get_runtime, LanceDBTable};
+use crate::error::{
+    handle_error, set_invalid_argument_message, set_unknown_error_message, LanceDBError,
+};
+use crate::types::{LanceDBMergeInsertConfig, LanceDBRecordBatchReader};
+
+/// Get table schema as Arrow C ABI
+///
+/// # Safety
+/// - `table` must be a valid pointer returned from `lancedb_connection_open_table`
+/// - `schema_out` must be a valid pointer to receive the Arrow schema
+/// - `error_message` can be NULL to ignore detailed error messages
+///
+/// # Returns
+/// - Error code indicating success or failure
+/// - The caller is responsible for releasing the schema using Arrow C ABI
+#[no_mangle]
+pub unsafe extern "C" fn lancedb_table_arrow_schema(
+    table: *const LanceDBTable,
+    schema_out: *mut *mut arrow_schema::ffi::FFI_ArrowSchema,
+    error_message: *mut *mut c_char,
+) -> LanceDBError {
+    if table.is_null() || schema_out.is_null() {
+        set_invalid_argument_message(error_message);
+        return LanceDBError::InvalidArgument;
+    }
+
+    let tbl = &(*table).inner;
+    let runtime = get_runtime();
+
+    match runtime.block_on(tbl.schema()) {
```

Comment:
I think `.block_on` can only be called once (might depend on the flavor of runtime you are using[^1]). So as written I don't think any of these functions could be called concurrently, which is probably an issue.

You should take a look what I did for pylance on the runtime: https://github.com/lancedb/lance/blob/main/python/src/executor.rs

That runs `block_on` in a background thread, and calling threads will use `spawn()` to run the futures. It also has nice handling for interrupts for cancellation, which I think it a worthwhile feature.

[^1]: https://docs.rs/tokio/latest/tokio/runtime/struct.Runtime.html#method.block_on